File: frontend/src/utils/sentry.ts
/**
 * Sentry Error Tracking Configuration
 *
 * Initializes Sentry for React error tracking.
 * DSN is provided via VITE_SENTRY_DSN environment variable.
 */
import * as Sentry from '@sentry/react';

export function initSentry() {
  const dsn = import.meta.env.VITE_SENTRY_DSN as string | undefined;

  if (!dsn) {
    console.log('[TradeUp] Sentry DSN not configured, error tracking disabled');
    return;
  }

  const environment = import.meta.env.MODE || 'development';

  Sentry.init({
    dsn,
    environment,
    release: `tradeup-frontend@${import.meta.env.VITE_APP_VERSION || '1.0.0'}`,

    // Performance monitoring
    tracesSampleRate: 0.1, // Sample 10% of transactions

    // Session replay (disabled for now to save quota)
    replaysSessionSampleRate: 0,
    replaysOnErrorSampleRate: 0.1, // Sample 10% of error sessions

    // Filter out noisy errors
    beforeSend(event, hint) {
      const error = hint.originalException;

      // Skip network errors (handled by axios interceptors)
      if (error instanceof TypeError && error.message.includes('fetch')) {
        return null;
      }

      // Skip cancelled requests
      if (error instanceof Error && error.name === 'CanceledError') {
        return null;
      }

      return event;
    },

    // Don't send personal info
    sendDefaultPii: false,

    // Ignore common browser extension errors
    ignoreErrors: [
      'ResizeObserver loop limit exceeded',
      'ResizeObserver loop completed with undelivered notifications',
      'Non-Error promise rejection captured',
    ],
  });

  console.log(`[TradeUp] Sentry initialized (env: ${environment})`);
}

/**
 * Capture an exception with optional context
 */
export function captureException(error: Error, context?: Record<string, unknown>) {
  Sentry.captureException(error, {
    extra: context,
  });
}

/**
 * Capture a message with optional level
 */
export function captureMessage(
  message: string,
  level: 'debug' | 'info' | 'warning' | 'error' | 'fatal' = 'info',
  context?: Record<string, unknown>
) {
  Sentry.captureMessage(message, {
    level,
    extra: context,
  });
}

/**
 * Set user context for Sentry events
 */
export function setUser(user: { id?: string; email?: string; shop?: string }) {
  Sentry.setUser(user);
}

/**
 * Clear user context (on logout)
 */
export function clearUser() {
  Sentry.setUser(null);
}

/**
 * Add breadcrumb for debugging
 */
export function addBreadcrumb(
  message: string,
  category: string,
  data?: Record<string, unknown>
) {
  Sentry.addBreadcrumb({
    message,
    category,
    data,
    level: 'info',
  });
}

// Export Sentry for advanced usage
export { Sentry };

File: app/utils/sentry.py
"""
Sentry Error Tracking Integration.

Captures exceptions and performance data for TradeUp.
"""
import os
import logging

logger = logging.getLogger(__name__)


def init_sentry(app=None):
    """
    Initialize Sentry error tracking.

    Sentry DSN should be set via SENTRY_DSN environment variable.
    If not set, Sentry will be disabled (development mode).

    Args:
        app: Optional Flask app for integrations
    """
    dsn = os.getenv('SENTRY_DSN')
    if not dsn:
        logger.info('Sentry DSN not configured, error tracking disabled')
        return None

    try:
        import sentry_sdk
        from sentry_sdk.integrations.flask import FlaskIntegration
        from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration

        environment = os.getenv('FLASK_ENV', 'development')
        release = os.getenv('RAILWAY_GIT_COMMIT_SHA', 'unknown')[:7]

        sentry_sdk.init(
            dsn=dsn,
            environment=environment,
            release=f"tradeup@{release}",
            integrations=[
                FlaskIntegration(transaction_style="url"),
                SqlalchemyIntegration(),
            ],
            # Performance monitoring
            traces_sample_rate=0.1,  # Sample 10% of transactions
            profiles_sample_rate=0.1,  # Profile 10% of sampled transactions

            # Filter out health check noise
            before_send=filter_events,

            # Don't send personal info
            send_default_pii=False,

            # Attach request data
            attach_stacktrace=True,
        )

        logger.info(f'Sentry initialized (env: {environment})')
        return sentry_sdk

    except ImportError:
        logger.info('sentry-sdk not installed, error tracking disabled')
        return None


def filter_events(event, hint):
    """
    Filter out noisy events before sending to Sentry.

    Returns None to drop the event, or the event to send it.
    """
    # Skip health check errors
    if 'request' in event:
        url = event['request'].get('url', '')
        if '/health' in url or url.endswith('/'):
            return None

    # Skip 404s
    if 'exception' in event:
        for exception in event['exception'].get('values', []):
            if exception.get('type') == 'NotFound':
                return None

    return event


def capture_exception(error, extra=None):
    """
    Capture an exception to Sentry.

    Args:
        error: The exception to capture
        extra: Optional dict of extra context
    """
    try:
        import sentry_sdk
        with sentry_sdk.push_scope() as scope:
            if extra:
                for key, value in extra.items():
                    scope.set_extra(key, value)
            sentry_sdk.capture_exception(error)
    except ImportError:
        pass


def capture_message(message, level='info', extra=None):
    """
    Send a message to Sentry.

    Args:
        message: The message to send
        level: Log level (debug, info, warning, error, fatal)
        extra: Optional dict of extra context
    """
    try:
        import sentry_sdk
        with sentry_sdk.push_scope() as scope:
            if extra:
                for key, value in extra.items():
                    scope.set_extra(key, value)
            sentry_sdk.capture_message(message, level=level)
    except ImportError:
        pass


def set_user(user_id=None, email=None, shop=None):
    """
    Set user context for Sentry events.

    Args:
        user_id: User/staff ID
        email: User email
        shop: Shopify shop domain
    """
    try:
        import sentry_sdk
        sentry_sdk.set_user({
            'id': user_id,
            'email': email,
            'shop': shop,
        })
    except ImportError:
        pass
